examples/http/client.cpp

```diff
@@ -0,0 +1,123 @@
+      .add<http::method>("method,m", "HTTP method to use")
+      .add<std::string>("payload,p", "Optional payload to send");
+    opt_group{custom_options_, "tls"} //
+      .add<std::string>("ca-file", "CA file for trusted servers");
+  }
+
+  caf::settings dump_content() const override {
+    auto result = actor_system_config::dump_content();
+    caf::put_missing(result, "method", default_method);
+    return result;
+  }
+};
+
+int caf_main(caf::actor_system& sys, const config& cfg) {
+  // Get URI from config (positional argument).
+  if (cfg.remainder.size() != 1) {
+    std::cerr << "*** expected mandatory positional argument: URL" << std::endl;
+    return EXIT_FAILURE;
+  }
+  uri resource;
+  if (auto err = parse(cfg.remainder[0], resource)) {
+    std::cerr << "*** failed to parse URI: " << to_string(err) << std::endl;
+    return EXIT_FAILURE;
+  }
+  auto ca_file = caf::get_as<std::string>(cfg, "tls.ca-file");
+  auto method = caf::get_or(cfg, "method", default_method);
+  auto payload = caf::get_or(cfg, "payload", ""sv);
+  auto result
+    = http::with(sys)
+        // Lazy load TLS when connecting to HTTPS endpoints.
+        .context_factory([ca_file]() {
+          return ssl::emplace_client(ssl::tls::v1_2)().and_then(
+            ssl::load_verify_file_if(ca_file));
+        })
+        // Connect to the address of the resource.
+        .connect(resource)
+        // If we don't succeed at first, try up to 5 times with 1s delay.
+        .retry_delay(1s)
+        .max_retry_count(5)
+        // After connecting, send a get request.
+        .add_header_field("User-Agent", "CAF-Client")
+        .request(method, payload);
+  if (!result) {
+    std::cerr << "*** Failed to initiate connection: "
+              << to_string(result.error()) << std::endl;
+    return EXIT_FAILURE;
+  }
+  sys.spawn([res = result->first](event_based_actor* self) {
+    res.bind_to(self).then(
+      [](const http::response& r) {
+        std::cout << "Server responded with HTTP "
+                  << detail::to_underlying(r.code()) << ": " << phrase(r.code())
+                  << std::endl;
+        std::cout << "Header fields:" << std::endl;
+        for (const auto& [key, value] : r.header_fields())
+          std::cout << "- " << key << ": " << value << std::endl;
+        if (r.body().empty())
+          return;
+        std::cout << "Payload:" << std::endl;
+        if (is_valid_utf8(r.body())) {
+          std::cout << to_string_view(r.body()) << std::endl;
+        } else {
+          // Print 8 bytes per row in hex.
+          auto bytes = r.body();
+          while (!bytes.empty()) {
+            auto row = bytes.subspan(0, 8);
+            for (auto byte : row)
+              printf("%02x", std::to_integer<int>(byte));
+            std::cout << std::endl;
+            if (bytes.size() < 8)
+              bytes = bytes.subspan(bytes.size());
+            else
+              bytes = bytes.subspan(8);
```

Comment:
I don't think this is correct. The first argument is the offset. To chop off the first 8 bytes, this should `bytes.subspan(8)`, shouldn't it? https://en.cppreference.com/w/cpp/container/span/subspan

Reply:
Oh boy...

> The behavior is undefined if either Offset or Count is out of range.

What a dreadful design. Well, in this case:

```suggestion
            bytes = bytes.subspan(std::min(bytes.size(), 8u));
```

Reply:
Sorry for comment-spamming. Since the API is so terrible, this means the `subspan(0, 8)` above is also unsafe.

Just split this into a utility lambda that returns the row (up to 8 bytes) and the remainder. Could use like this:

```c++
const_byte_span row;
std::tie(row, byte) = splitAt(bytes, 8);
```

Reply:
No problem, I'm glad this is the only critique :smile:. I figured the first overload out, but yeah the first one also had the bug.

Reply:
> I'm glad this is the only critique

Erm... this was just a comment on something that I've noted before starting the actual review. 😅